### ISSUE TITLE
Grab-strike overcompensates for player velocity

### ISSUE THREAD

**Issue by SolraBizna** [endless-sky/endless-sky#3686]
I'm flying a souped-up Sojourner with Planetary-class engines and mainly Grab-Strikes for armament. If I am stationary, the Grab-Strike does a reasonable job of aiming at any ship I'm fighting, fast or slow. However, the faster I am moving, the worse its aim is. Unless the ship is directly behind me on a parallel course, the Grab-Strike effectively never hits its target, instead passing well in front. It seems as if the faster I am moving, the more the turret aiming code underestimates the Grab-Strikes' actual velocity. (Grab-Strikes do move "faster" when fired rearward than they do forward, because they have acceleration, drag, and homing.)

**Comment by kyzentun** [comment]
Here's my analysis as someone just glancing through the code from the outside.

In AI.cpp, function AI::AimTurrets, there's the loop for finding the enemy each turret can aim at.  In there is this code:
```c++
				// Only take the ship's velocity into account if this weapon
				// does not have its own acceleration.
				if(!weapon->Acceleration())
					v -= ship.Velocity();
```
The grab-strike has acceleration and drag, so ship velocity is ignored.  The Projectile constructor in Projectile.cpp uses the ship velocity.  Thus, the aiming calculations are wrong about the velocity of the projectile.  Since the grab strike only has a turn of .2, lower than any missile, it can't compensate the way missiles do.

If the ship velocity were used in the calculations, they would be wrong in a different way.  AI::RendezvousTime calculates when the projectile would reach the target.  But it works with a constant velocity, it won't calculate the right time for a weapon that has drag and acceleration.

So turreted accelerating weapons like the grab strike need a separate function for calculating time to impact that factors in their drag and acceleration.

**Comment by NCGThompson** [comment]
I am attempting to fix this using the method proposed by @kyzentun :

> So turreted accelerating weapons like the grab strike need a separate function for calculating time to impact that factors in their drag and acceleration.

 Where in game can a Grab-Strike or something else good for testing this be found?
Disclamer: I never looked at C++ before and have no idea what I'm doing so don't get your hopes to high.

**Comment by tehhowch** [comment]
If you are testing code changes, it is wise to

1) avoid using a "real" save
2) give yourself whatever is needed to evaluate results

So just make a system in which AI ships with Grab-Strikes spawn, give your own ship Grab-Strikes, and spawn yourself there. Tweak, recompile, and relaunch as needed.

**Comment by NCGThompson** [comment]
The current algorithm for finding the angle is derived from the equation where t is predicted rendezvous time and Θ is angle fired.
target position + (target velocity * t) = Θmuzzle speed * t
As explained [here.](https://www.gamasutra.com/blogs/KainShin/20090515/83954/Predictive_Aim_Mathematics_for_AI_Targeting.php)
we can find an algorithm for t as a function of absolute target position/velocity and muzzle speed:
```CPP
double a = v.Dot(v) - vp * vp;
double b = 2. * p.Dot(v);
double c = p.Dot(p);
double discriminant = b * b - 4 * a * c;
if(discriminant < 0.)
	return numeric_limits<double>::quiet_NaN();

discriminant = sqrt(discriminant);

// The solutions are b +- discriminant.
// But it's not a solution if it's negative.
double r1 = (-b + discriminant) / (2. * a);
double r2 = (-b - discriminant) / (2. * a);
```
We then select the lowest non-negative t, and use it to predict final position of the target. From there it's just SohCahToa.

What we want to do is modify this to account for the velocity of the gun. Fortunately the "force" of drag is directly proportional to velocity which means we can exploit a cool property of the the relationship of * and + that means we can divide a vector into pieces, apply drag to both, then add them.
All we have to do is get an expression for the offset:
Turret velocity * (drag coefficient^t + 1)/(2 - drag coefficient - t)
account for that finding Θ from t:
point at target position + target velocity * time - offset
and insert it into the previously mentioned equation:
target position + (target velocity * t) = Θmuzzle speed * t + offset

There is one catch. I consulted an engineer and he claimed that that last equation is "definitely unsolvable by hand."

What else can we try?
- Solve the equation not "by hand." We only need one person to find the solution once!
  - We can try using sample values and look for patterns in the results.
  - Guess and check (but it might be a poor use of time)
- Start from scratch with a totally different approach.
  - How can we get around solving for t?
- Use heuristic algorithm. Or a *more* heuristic algorithm.
  - Piecewise might work
  - Brief iterative approximation (also in the link above)

**Comment by endless-sky** [comment]
It would probably be sufficient to just not take the firing ship's velocity into account, since the result of the tracking behavior is that after the first few frames, the projectile's velocity is no longer affected by the firing ship's velocity.

But, I'm not sure it makes sense to make the Grab-Strike even more effective at tracking. The relative ineffectiveness at short ranges and high relative velocities is part of what balances out its long range and ability to fire a tracking projectile that consumes no ammunition.

**Comment by KiLEdEnNis** [comment]
I'd prefer giving it poor tracking or a high inaccuracy then, instead of sloppy AI(?)

**Comment by SolraBizna** [comment]
The Grab-Strike being ineffective under certain circumstances makes an interesting tradeoff against its otherwise high relative effectiveness. I just want it to be *explicit* rather than due to a bug.

We could add an attribute that controls the extent to which the rendezvous is miscalculated. Set that to 1 for the Grab-Strike to have it continue to behave exactly the same way, while mod-added weapons might want 0 for maximum accuracy, or something in between for partial compensation, or even something out of the 0–1 range for over-/anti-compensation.

**Comment by SolraBizna** [comment]
> My opinion on the effect on balance for grab strikes: Balance is the highest priority but to a player making trick shots is very satisfying. On the opposite end having ones turrets miss in the same way every time and not being able to fix it is very frustrating. Using random inaccuracy (by increasing the currently implemented attribute) is much better than predictable inaccuracy.

I agree and disagree. Having the grab-strike consistently fail to hit attackers while I was doing a Monty Python was very frustrating, true. On the other hand, it also meant I had to switch up my tactics, and ended up doing things that would have been sub-optimal for a more conventional turret-based setup... or just trying to line up my attacker with my retrograde vector. It's also nice to have a semi-unique weakness to exploit when *attacking* a grab-strike equipped ship.

Then again, it never really stopped being *frustrating*...
